jellyfin-core/src/main/kotlin/org/jellyfin/sdk/discovery/RecommendedServerDiscovery.kt

```diff
@@ -27,103 +34,100 @@ public class RecommendedServerDiscovery(
-
-		@Suppress("TooGenericExceptionCaught")
-		val info = try {
-			api.getPublicSystemInfo()
-		} catch (err: ConnectException) {
-			logger.debug("Could not connect to $address", err)
-			null
-		} catch (err: Exception) {
-			logger.error("Could not retrieve public system info for $address", err)
-			null
+		val info: Response<PublicSystemInfo>?
+		val responseTime = measureTimeMillis {
+			@Suppress("TooGenericExceptionCaught")
+			info = try {
+				api.getPublicSystemInfo()
+			} catch (err: TimeoutException) {
+				logger.debug("Could not connect to $address", err)
+				null
+			} catch (err: InvalidStatusException) {
+				logger.debug("Received unexpected status ${err.status} from $address", err)
+				null
+			} catch (err: InvalidContentException) {
+				logger.debug("Could not parse response from $address", err)
+				null
+			} catch (err: Throwable) {
+				logger.error("Could not retrieve public system info for $address", err)
+				null
+			}
 		}
-		val endTime = System.currentTimeMillis()
 
 		return SystemInfoResult(
 			address = address,
 			systemInfo = if (info != null && info.status == HTTP_OK) info.content else null,
-			responseTime = endTime - startTime,
+			responseTime = responseTime,
 		)
 	}
 
-	@Suppress("MagicNumber")
-	private fun assignScore(result: SystemInfoResult, parentResult: SystemInfoResult? = null): RecommendedServerInfo {
-		var points = 0
-
-		// Security
-		if (result.address.startsWith(HTTPS_PREFIX)) points += 3
-
-		// Speed
-		when {
-			result.responseTime < 500 -> points += 3
-			result.responseTime < 1500 -> points += 2
-			result.responseTime < 5000 -> points += 1
-		}
-
-		// Compatibility
-		val version = result.systemInfo?.version?.let(ServerVersion::fromString)
-		if (version != null) {
-			if (version >= Jellyfin.apiVersion) points += 1
-			if (version >= Jellyfin.minimumVersion) points += 1
-		}
-
-		val productName = result.systemInfo?.productName
-		if (productName != null && !productName.equals(PRODUCT_NAME, ignoreCase = true)) points = 0
-
-		// Minimum amount of points: 0
-		// Maximum amount of points: 8
-		val score = when {
-			points < 3 -> RecommendedServerInfoScore.BAD
-			points < 6 -> RecommendedServerInfoScore.OK
-			else -> RecommendedServerInfoScore.GOOD
-		}
-
-		return RecommendedServerInfo(result.address, result.responseTime, score, result.systemInfo, parentResult?.address)
+	/**
+	 * Discover all servers in the [servers] flow and retrieve the public system information to assign a score.
+	 * Returned servers are not ordered by score. Use [minimumScore] to automatically remove bad matches.
+	 */
+	public suspend fun discover(
+		servers: Flow<String>,
+		minimumScore: RecommendedServerInfoScore,
+	): Flow<RecommendedServerInfo> = withContext(Dispatchers.IO) {
+		servers
+			.map(::getSystemInfoResult)
+			.map(::assignScore)
+			.filter { serverInfo ->
+				// Use [minimumScore] to filter out bad score matches
+				serverInfo.score.score >= minimumScore.score
+			}
 	}
```

Comment:
Should use [flowOn](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/flow-on.html) (I don't even know if withContext works here).

Reply:
https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/flow-on.html

The withContext works, flowOn can be used to run a part of the flow in a different context.

Reply:
lmao, I'm blind. I linked the exact same page. Didn't see the `withContext` in the example.

Reply:
I'm blind too because I did not notice the link in your reply 😂